### DISCUSSION TITLE
In App Purchases free 1 issue

### DISCUSSION THREAD

**Discussion by VAROME** [vfsfitvnm/frida-il2cpp-bridge#377] [Q&A]
i am trying on a game named Frontier Wars 

```

    const update = Tizen_IAPClass.method("Buy")

    update.implementation = function(
this: Il2CppThis,
      stuck:any
    ) {
    
        const instance = this
        let spawnCount = 0;
 
 
 
        Tizen_IAPClass.fields.forEach((field => 
    {
      BuySuccessMethod.implementation = function(this: Il2Cpp.Object | Il2Cpp.Class
        ,param1:any
         ) {

     if (spawnCount < 5) {

          param1 = "com.ffs.frontier.war.6000diamonds";
        this.method("BuySuccess").invoke(param1); 
        console.log("tr 1y");
 
          spawnCount++;
           }  };
         }
      )
      )
    ;
    
 
        this.method<void>("Buy").invoke(stuck)
    
    } 
    
```

i hope you help me to make a IAP free whetther it is this way or another way you have

**Comment by VAROME** [comment]
Edit: the error in the log is 2 things
```
a thunk
il2cpp: couldn't set implementation for method BuySuccess as it has already been replaced by 
a thunk
il2cpp: couldn't set implementation for method BuySuccess as it has already been replaced by 
a thunk
il2cpp: couldn't set implementation for method BuySuccess as it has already been replaced by 
a thunk
il2cpp: couldn't set implementation for method BuySuccess as it has already been replaced by 
a thunk```
if i fix it 
i get instead this 
```
il2cpp: cannot invoke non-static method BuySuccess as it must be invoked throught a Il2Cpp.Object, not a Il2Cpp.Class
    at invoke (il2cpp/structs/method.ts:215)
    at <anonymous> (src/index.ts:226)
    at forEach (native)
    at <anonymous> (src/index.ts:230)
    at call (native)
    at <anonymous> (il2cpp/structs/method.ts:334)
    ```
    i know already hooking and tested a lot , using update, and non using it , static and non static with and without paramters , all successfull , but this, nope . 
and if there is a better way of calling functions without modding them would be great

**Reply by christiancoleman** [reply]
With respect to the bit about not calling it as a static method. Had you already tried instantiating it like this:
`const instance = Tizen_IAPClas.new()`
I think if I’m understanding you then you did and that’s when you’re seeing the thunk error messages. Correct me if I’m wrong though. I’m having trouble replacing a function myself.

**Reply by DragonRiseo** [reply]
i have not tried anything about it yet , but i upgraded my knowledge since then  and now i can gather all instances to any class and use those objects to hook anything , 
so i may give it a try again later

**Reply by DragonRiseo** [reply]
as for the thunk  , i now just revert it

**Reply by christiancoleman** [reply]
I'd very much be interested in seeing some sample code. Is it by chance doing something like this?

```
var classInstance: Il2Cpp.Object = null;
Il2Cpp.GC.choose(someClass).forEach((instance: Il2Cpp.Object) => {
		classInstance = instance;
});
console.log(classInstance);	
```

 I'm still having issues replacing a function. I thought if I grabbed an active instance it would work out but alas I'm still coming up short.

**Reply by DragonRiseo** [reply]
no , 
no null objects , it would make some stuff not work well , 
you can just add an if statement to check if it is not null then act

**Reply by DragonRiseo** [reply]
and for heap scanning dont use gc.choose , it is shit especially when you have a menu, it wont stop easy by turning the switch off

**Reply by christiancoleman** [reply]
hate to beat a dead horse, but would you mind dropping a snippet on how you grab an active instance? would appreciate it! <3

**Reply by christiancoleman** [reply]
Hey @VAROME any possibility you could post that code snippet? Could have swore you replied here that you would when ya got home but it seems to have disappeared.

**Reply by VAROME** [reply]
will do
heap scanning use the second method not GC 
https://github.com/vfsfitvnm/frida-il2cpp-bridge/wiki/Snippets#heap-scanning
 
 will help you better with your code especially with something like a toggle , 
 and then just call it with setInterval when you test or debug , cause it gets called once [or if you hook a class and gather its objects will be called whenever theclass is called makiing the best lag you would ever see ]so you call it setinterval  and set it to 3 seconds [3000 cause it supports milii directly ] 
 
```
const SystemType = assemblycSharp.class("system.type"); 

  const snapshot = Il2Cpp.MemorySnapshot.capture();
    snapshot.objects.filter(Il2Cpp.isExactly(SystemType)).forEach((instance: Il2Cpp.Object) => {
// then do something 
    });
    snapshot.free();
```
encapsulate it inside a button 
```
const button = menu.button.type.OnOffButton ; 
button ("LOL TEST", Function => {
 // some shit 

NewHookFridaVFS  someField8 = hook.function.nestedClass("_playerHP").field("_currHp"); 

console.log(
someField8.value
);
 
someField8.value :any = someField8.value *VRSliderValue;
 
ToastVF.toast("DAMAGE MULTIPLAYER LOL XD").short(0) ; 



 } 

```

**Reply by VAROME** [reply]
i have edited the comment multiple times so you see the codes , will edit it again

**Reply by VAROME** [reply]
@christiancoleman you stll follow ?

**Reply by christiancoleman** [reply]
@VAROME yep I'm here and viewing it. Thanks man! The heap scanning using a memory snapshot definitely makes way more sense than using the garbage collector.

**Comment by VAROME** [comment]
anything ?

**Comment by VAROME** [comment]
2 weeks and nothing

**Comment by VAROME** [comment]
this is for something like  @christiancoleman 

![1](https://github.com/vfsfitvnm/frida-il2cpp-bridge/assets/135765220/8933fd63-6ced-4c3b-9e36-0a970daee938)


will do
heap scanning use the second method not GC 
https://github.com/vfsfitvnm/frida-il2cpp-bridge/wiki/Snippets#heap-scanning
 
 will help you better with your code especially with something like a toggle , 
 and then just call it with setInterval when you test or debug , cause it gets called once [or if you hook a class and gather its objects will be called whenever theclass is called makiing the best lag you would ever see ]so you call it setinterval  and set it to 3 seconds [3000 cause it supports milii directly ] 
 
```
const SystemType = assemblycSharp.class("system.type"); 

  const snapshot = Il2Cpp.MemorySnapshot.capture();
    snapshot.objects.filter(Il2Cpp.isExactly(SystemType)).forEach((instance: Il2Cpp.Object) => {
// then do something 
    });
    snapshot.free();
```
encapsulate it inside a button 
```
const button = menu.button.type.OnOffButton ; 
button ("LOL TEST", Function => {
 // some shit 

NewHookFridaVFS  someField8 = hook.function.nestedClass("_playerHP").field("_currHp"); 

console.log(
someField8.value
);
 
someField8.value :any = someField8.value *VRSliderValue;
 
ToastVF.toast("DAMAGE MULTIPLAYER LOL XD").short(0) ; 



 } 

```

**Reply by christiancoleman** [reply]
Oh I didn't even see this reply. This makes sense! I really appreciate you replying back to me. Good looking out!

**Reply by VAROME** [reply]
using this 
https://github.com/maarsalien/frida-android-mod-menu
https://github.com/commonuserlol/frida-java-menu
the template makes a menu for frida , like LGLTEAM , 
the advantage here is that in LGL you can only hook things that are normal like int, bool, long, string 
but using frida you can directly hook what you want , even if it is SOMESHIT literally , you just tell it : it  is a method or field named "something" and it is located in class / nested class "something" and just tell it to hook , and return what you want , the type of the field or the method aint even needed 
that  is why for example lgl you need to use Update offset of the class to hook a field or creating a pointer for a method , but using ths you can just hook it once , and create a button to call it directly when clicked , or directlry hook the field , without offset , or even update ,  
the only thing left to be created is fps hacks like aimbot or esp [esp is already created by maarsalien but not the best]

**Comment by RevealedSoulEven** [comment]
#459

**Comment by IrvanWijayaSardam** [comment]
looks like my question is belongs to here, i hope somebody can help, i still learning, how do i call <Awake>b__25_7 ? i've got this message 

`il2cpp: cannot invoke non-static method <Awake>b__25_7 as it must be invoked throught a Il2Cpp.Object, not a Il2Cpp.Class
    at invoke (il2cpp/structs/method.ts:214)
    at <anonymous> (src/index.ts:30)
    at call (native)
    at <anonymous> (il2cpp/structs/method.ts:328)`
    
    
    this is my frida script
    `import "frida-il2cpp-bridge";

console.log("Frida loaded successfully !!");

Il2Cpp.perform(function () {
        

    const AssemblyCSharp = Il2Cpp.domain.assembly("Assembly-CSharp").image;

    const TableManager = AssemblyCSharp.class("TableManager");
    const LobbyMenu = AssemblyCSharp.class("LobbyMenu");

    const OnBackBtn = TableManager.method("OnBackBtn");
    const AwakeB257 = LobbyMenu.method("<Awake>b__25_7");

    OnBackBtn.implementation = function (unused) {
        console.log("Back Button Hooked !!");
        AwakeB257.invoke();
    };

});
`

and this is the awake function from dnSpy
`	private void <Awake>b__25_7() {}
`
